### PR TITLE
Initial template support

### LINKED ISSUE #103: Fitting handlebars template engine into Response/Serializer or Middleware system
I was looking into tower-web to integrate handlebars into its system.

The typical usage of handlebars in web application is to create a server-scoped registry, that holds all templates in memory. We use the registry to render templates returned from handler functions. 

The ideal API for tower-web in my mind is to create a `Template` struct, which contains template `name` and template `data`. The `Template` struct will implement `Response` to render the data structure into actual http response. The problem is `Response` trait only provides a request-scoped `Context` so there seems to be no way to access server-scoped handlebars registry.

So is it possible to bind the registry into some data structure that has server-scoped lifetime? And be accessible from `Response` trait and middlewares.

### PR DESCRIPTION
**Pull request by carllerche**
## Overview

This PR primarily provides a first stab at HTML templates with Tower Web. Initially, handlebars will be a dependency of tower-web, but the integration is done without much coupling. Other template engines will be able to be plugged in.

To support templates, this PR also adds initial support for user provided response serializers. The same mechanism used for custom template libs can be used to add support for things like XML responses.

[Example](https://github.com/carllerche/tower-web/blob/handlebars-support/examples/html_handlebars.rs) of defining a response that uses a [template](https://github.com/carllerche/tower-web/blob/handlebars-support/templates/foo.hbs) to render.

## Notes

* `Serializer` now takes a `SerializerContext` argument. This argument contains contextual information like the `template` annotation specified on the derived response type. It also provides access to the resource module, type name, and handler name. This can be used to use a conventional path for the template.

* `response::Context` also has the resource module, resource name, and handler name.

* `Response::into_http` now returns a `Result`. This allows for better error handling.

Closes #103

### CODE REVIEW

File: tests/html_handlebars.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+use support::*;
+
+use handlebars::Handlebars as Registry;
+use tower_web::view::Handlebars;
+
+#[derive(Debug)]
+struct TestHandlebars;
+
+#[derive(Response, Debug)]
+#[web(template = "foo")]
```
Comment (shepmaster) [inline tests/html_handlebars.rs]: Seems dicey to rely on the same template in two different contexts.

File: tests/html_handlebars.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+extern crate handlebars;
+extern crate http;
+#[macro_use]
+extern crate tower_web;
+
+#[macro_use]
+mod support;
+use support::*;
+
+use handlebars::Handlebars as Registry;
```
Comment (shepmaster) [inline tests/html_handlebars.rs]: Now I'm conflicted... should this be consistent?

File: src/view/mod.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+//! Render content using templates
+//!
+//! Current, Handlebars is the only supported template engine. In time, the
```
Comment (shepmaster) [inline src/view/mod.rs]: typo: "currently"

File: src/view/handlebars.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    return Ok(rendered.into());
+                }
+                Err(err) => {
+                    error!("error rendering template; err={:?}", err);
+                    return Err(::error::ErrorKind::internal().into())
+                }
+            }
+        }
+
+        // TODO: Add conventional template lookup
```
Comment (shepmaster) [inline src/view/handlebars.rs]: What does "conventional" mean in this context?

File: src/view/handlebars.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    ///
+    /// The serializer renders handlebar templates using the response value to
+    /// populate template variables. The response value must have
+    /// `#[derive(Response)]` and a template name specified using the
+    /// `#[web(template = "<template name>")]`.
+    ///
+    /// Templates are loaded from the `templates` directory in the crate root
+    /// and have the `.hbs` file extension.
+    pub fn new() -> Handlebars {
+        let mut registery = Registery::new();
```
Comment (shepmaster) [inline src/view/handlebars.rs]: Ugh, you have got to be kidding me. The typo is in the library?

File: tests/html_handlebars.rs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+fn hb() -> Handlebars {
+    use std::env;
+    use std::path::Path;
+
+    let mut registry = Registry::new();
+
+    let dir = Path::new(&env::var("CARGO_MANIFEST_DIR").unwrap())
+        .join("tests/templates/hbs");
+
+    registry.register_templates_directory(".hbs", dir).unwrap();
```
Comment (shepmaster) [inline tests/html_handlebars.rs]: Ah, it's not the same? Picking a different name from `foo` might be a good call then.

File: tests/html_handlebars.rs (hunk 2 above)
Comment (carllerche) [reply tests/html_handlebars.rs]: `handlebars::Handlebars` is actually a re-export of `Registry` internally :shrug:

File: src/view/handlebars.rs (hunk 5 above)
Comment (carllerche) [reply src/view/handlebars.rs]: Yeah no... all me. My keyboard loved to add `e`s everywhere and I use vim's autocomplete a lot... so, pretty much fail. (Also, I shipped off my laptop to have the keyboard repaired finally).

**Comment by carllerche** [comment]
@shepmaster Having the template specified on the action makes sense for *most* cases. I think you need to be able to also be able to specify the template on the response struct as well for cases like the response type is `either` (i.e. you need to pick the template at runtime).

**Comment by carllerche** [comment]
@shepmaster also, thanks for the review! Super valuable feedback. I'll work on the suggestions.

File: examples/html_handlebars.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+/// The type is annotated with `#[derive(Response)]`, this allows `MyResponse`
+/// to be used as a response to resource methods.
+///
+/// We are using the handlebars serializer to render the HTML response. It
+/// requires that a template to render is specified. This is done with the
+/// `#[web(template = "<template name>")]` attribute.
+///
+/// The default location to look for templates is in the `templates` directory
+/// at the crate root. To make the example work, this crate has a handlebars
+/// template at "foo.hbs" in the templates directory at the crate root.
```
[the remark this answers is not in the document]
Comment (carllerche) [reply examples/html_handlebars.rs]: I believe that there can be subdirectories, but no subdirectory is picked by convention yet. You would need to specify the subdirectory in the template name.

I would like a more conventional approach, but as far as I could tell, it was difficult to make work.

**Comment by carllerche** [comment]
@shepmaster I added handler level `template` support [here](https://github.com/carllerche/tower-web/pull/115/commits/49898e314d3f7be9475ff3e7cc9ba58b1aab0bcb#diff-2bbbec5968b025460fda919b11bbb1faR54).

I need to add more tests and update examples still...

File: src/view/handlebars.rs (hunk 4 above)
Comment (carllerche) [reply src/view/handlebars.rs]: Oh, that means pick a template name using a convention (like the mod / handler name).

**Pull request merged**